### PR TITLE
fix(scripts): scope guard turn_base across history rewrites

### PR DESCRIPTION
**Pull request by vanyastaff**
## Problem

`stop-gate.sh` and `intent-gate.sh` derive the current turn's touched-crate / touched-file set from `turn_base..HEAD`, where `turn_base` is a fixed commit SHA captured at turn start (`turn-reset.sh` A0).

After the working branch is rebased (`git rebase --onto origin/main <old-base> <branch>`) or its upstream base is squash-merged into `main` and the branch is reparented, `turn_base` points at a commit on an **abandoned history line**. The `turn_base..HEAD` diff (and the working-tree/staged/untracked union both hooks build) then spans the **entire delta between the old and new base** — i.e. the whole squash-merged upstream PR — so crates that were only changed by that already-merged work are reported as "changed this turn". The Stop-gate then demands a fresh `cargo clippy -p <crate> -- -D warnings` + `cargo nextest run -p <crate>` for each (minutes per crate), training the operator to treat the guard as noise. `intent-gate.sh` showed the identical stale-base churn (spurious net-LoC / new-file / blob counting).

## Fix

New `effective_turn_base()` in `_lib.sh`, called by both hooks before the committed-this-turn diff arm:

- **`turn_base` still an ancestor of `HEAD`** (the common, no-rewrite case) → returned unchanged. **Zero behavior change.**
- **Not an ancestor** (`git merge-base --is-ancestor` fails ⇒ history rewrite) → repin to the branch's `merge-base` with the first of `origin/main` / `main` / `@{upstream}` that resolves. That point is **after** any squash-merge, so the merged upstream work drops out, while commits made on *this* branch are still seen.
- **No upstream ref resolves** → `""`; callers skip the diff arm — documented safe degradation, identical to the unborn-branch path. B-union (`impl_files_edited`) + working-tree `git status` ground truth still enforce the green gate.

Only the touched-set **scoping** changes. The D10 no-cheat guarantee (edit-guard B + clean-gate recorder A2 + Stop-gate C + lefthook/CI) is untouched — `record.sh` / `edit-guard.sh` unmodified; the `// budget-justified:` / escape semantics of `intent-gate.sh` are preserved.

### Options considered

| # | Option | Verdict |
|---|--------|---------|
| 1 | Always `origin/main...HEAD` | Rejected — regresses the common multi-turn case (re-demands every prior turn's committed crates; `gate_green` is reset per turn) |
| **2** | **Repin when `turn_base` no longer ancestor of HEAD** | **Chosen — exact prior semantics in the no-rewrite case, graceful recovery on rewrite, minimal** |
| 3 | Date-window commits + content-equality vs `origin/main` | Rejected — rebase rewrites commit-dates; per-file blob compare is complex, violates "minimal" |
| 4 | Memoize A2 green per crate by tree hash | Rejected — does not fix root cause (a rebased crate's tree *does* change → memo misses); also mutates the A2 core |

## Verification

- `task hooks:test` → `ALL GUARD TESTS PASSED` (exit 0). New regression cases, all green:
  - `effective_turn_base` unit: ancestor passthrough / empty / repins-stale-base
  - **`C ignores rebased-in crate`** — the reproduced false positive (RED → GREEN)
  - **`C still blocks real post-rebase crate`** — genuine no-cheat block still fires when the real this-turn crate has no recorded green gate
  - **`E ignores rebased-in net-LoC`** — intent-gate no longer counts the rebased-in upstream delta
- Manual rebased fixture: `effective_turn_base` repins ≠ stale base; `stop-gate.sh` / `intent-gate.sh` exit `0` on a 600-LoC rebased-in upstream, `stop-gate.sh` exits `2` when the real crate is ungreen.
- TDD: tests written first, watched fail with the predicted reason, then implemented, watched pass.
- lefthook pre-commit (typos + convco) and pre-push (full workspace gate) passed; no hooks bypassed.

## Scope

`.claude/hooks/_lib.sh` (+34), `stop-gate.sh` (+4/-1), `intent-gate.sh` (+4/-1), `test/run.sh` (+75 regression). No unrelated hooks changed.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Improved state tracking for code changes during branch operations

* **Improvements**
  * Enhanced handling of rebased and squash-merged branches
  * Better recovery and baseline computation during complex git operations

* **Tests**
  * Added test coverage for rebase and upstream merge scenarios
  * Extended regression testing for state management edge cases

[Review Change Stack](https://app.coderabbit.ai/change-stack/vanyastaff/nebula/pull/726?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: .claude/hooks/_lib.sh [hunk 1]
```diff
@@ -29,6 +29,40 @@ load_state() { # $1=path -> always {impl_files_edited:[...],gate_green:[...],tur
...
+effective_turn_base() {
+  local cwd="$1" tb="$2" ref mb
+  [ -n "$tb" ] || { printf ''; return 0; }
+  if git -C "$cwd" merge-base --is-ancestor "$tb" HEAD 2>/dev/null; then
+    printf '%s' "$tb"; return 0          # no rewrite — prior semantics intact
+  fi
+  for ref in origin/main main '@{upstream}'; do
+    git -C "$cwd" rev-parse --verify -q "$ref" >/dev/null 2>&1 || continue
+    mb="$(git -C "$cwd" merge-base HEAD "$ref" 2>/dev/null)" || continue
+    [ -n "$mb" ] && { printf '%s' "$mb"; return 0; }
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply .claude/hooks/_lib.sh]: Fixed in f59cec7c. A0 (`turn-reset.sh`) now snapshots `turn_base_patch_ids` — the `git patch-id --stable` of every pre-turn branch commit (`upstream-mb..HEAD` at turn start). When `turn_base` is orphaned by `git rebase --onto`, `effective_turn_base` walks the rebased line and matches those patch-ids to recover the rewritten counterpart of the original `turn_base`; the diff arm scopes to commits ABOVE that on the new line = this turn only, so pre-turn commits stay outside the touched set. Falls back to upstream merge-base when patch-id match fails (rebase with conflicts changed the diff). Regression added covering your reproducer: `C ignores pre-turn branch commit (P1)` + `C still blocks real this-turn crate (P1)` (no over-relax) + the intent-gate analogue.
